File: versions/cpp_combinator/benchmark.cpp
#include <iostream>
#include <fstream>
#include <chrono>
#include <sstream>
#include "lexer.cpp"

int main(int argc, char** argv) {
    if (argc < 3) {
        std::cerr << "Usage: " << argv[0] << " <input_file> <iterations>\n";
        return 1;
    }

    std::ifstream t(argv[1]);
    if (!t.is_open()) return 1;
    std::stringstream buffer;
    buffer << t.rdbuf();
    std::string source = buffer.str();

    int iterations = std::atoi(argv[2]);
    long tokens_count = 0;

    auto start_time = std::chrono::high_resolution_clock::now();

    for (int i = 0; i < iterations; i++) {
        CloudPolLexer lexer(source);
        while (true) {
            Token tok = lexer.nextToken();
            if (tok.code == TokenCode::END_OF_FILE) break;
            tokens_count++;
        }
    }

    auto end_time = std::chrono::high_resolution_clock::now();
    long duration_ms = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time).count();

    std::cout << "Native Combinator (C++) Parsed " << tokens_count << " tokens in " << iterations << " iterations.\n";
    std::cout << "Time taken: " << duration_ms << " ms\n";

    return 0;
}

File: versions/antlr/Benchmark.java
import org.antlr.v4.runtime.*;
import java.nio.file.Files;
import java.nio.file.Paths;

public class Benchmark {
    public static void main(String[] args) throws Exception {
        if (args.length < 2) {
            System.err.println("Usage: java Benchmark <input_file> <iterations>");
            System.exit(1);
        }
        
        String source = new String(Files.readAllBytes(Paths.get(args[0])));
        int iterations = Integer.parseInt(args[1]);
        long tokensCount = 0;
        
        long startTime = System.currentTimeMillis();
        
        for (int i = 0; i < iterations; i++) {
            CloudPolLexer lexer = new CloudPolLexer(CharStreams.fromString(source));
            while (true) {
                Token tok = lexer.nextToken();
                if (tok.getType() == Token.EOF) break;
                tokensCount++;
            }
        }
        
        long endTime = System.currentTimeMillis();
        
        System.out.println("ANTLR (Java) Parsed " + tokensCount + " tokens in " + iterations + " iterations.");
        System.out.println("Time taken: " + (endTime - startTime) + " ms");
    }
}
